File: 1287.element-appearing-more-than-25-in-sorted-array.cpp
/*
 * @lc app=leetcode id=1287 lang=cpp
 *
 * [1287] Element Appearing More Than 25% In Sorted Array
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int findSpecialInteger(vector<int>& arr) {
    int limit = arr.size() / 4 + 1;
    int last = -1;
    int count = 1;
    for (const auto num : arr) {
      if (last != num) {
        last = num;
        count = 1;
        continue;
      }
      if (++count >= limit) {
        return num;
      }
    }
    return arr[0];
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> arr;
  arr = {1, 2, 2, 6, 6, 6, 6, 7, 10};
  ASSERT_EQ(s.findSpecialInteger(arr), 6);
  arr = {1, 1};
  ASSERT_EQ(s.findSpecialInteger(arr), 1);
  arr = {1};
  ASSERT_EQ(s.findSpecialInteger(arr), 1);
  arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 12, 12, 12};
  ASSERT_EQ(s.findSpecialInteger(arr), 12);
  return 0;
}

File: 1287.element-appearing-more-than-25-in-sorted-array.rs
/*
 * @lc app=leetcode id=1287 lang=rust
 *
 * [1287] Element Appearing More Than 25% In Sorted Array
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn find_special_integer(arr: Vec<i32>) -> i32 {
        0
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::find_special_integer(vec![1, 2, 2, 6, 6, 6, 6, 7, 10]),
        6
    );
}

#[test]
fn test2() {
    assert_eq!(Solution::find_special_integer(vec![1, 1]), 1);
}
